Add FidelityFX Super Resolution 2.2 (FSR 2.2.1) support.

DarioSamo opened the pull request:
# FidelityFX Super Resolution 2.2

Introduces support for FSR2 as a new upscaler option available from the project settings.

It's worth noting that while I've done my best effort in making improvements to the pipeline to provide the information as well as possible to FSR2 with the minimal performance impact, **some artifacts inherent to the algorithm are to be expected**. This is a pretty competitive area at the moment where both DLSS and FSR aren't perfect in all cases and it'll vary a lot on the type of content that is used. While I can address artifacts introduced by my own implementation due to errors, if the contents of the render fit what FSR2 expects, there's not much else I can do. Therefore, we should probably limit reporting to noticeable issues (e.g. broken results) that aren't FSR2 problems.

## To verify

There's some UX issues where it'd be appreciated to gather some feedback on how they should be solved.

- We can arguably improve loading times in general if FSR2 is not used by opting to move shader initialization outside of the FSR2Effect constructor, but maybe the delay when turning on the feature is not desired.

While the FSR version included in the PR mostly matches the public GitHub version, some changes have been made to better accommodate for Godot's needs.

## TODO

- [x] MSAA seems to work with TAA but not FSR2. Probably not desired anyway but should be fixable.
- [x] Mark modified sections of FSR2 code with comment blocks.
- [x] Viewport constant for FSR2 is missing or not exposed
- [x] Extra newlines on shader files
- [x] Move FSR2 Context to forward clustered render buffers
- [x] Add extra `get_texture_slice_view` function for public-facing API
- [x] Fix validation errors when using SSS when using SSS or other separate specular effects.
- [x] Clear alpha to zero after opaque pass when using SSS or other separate specular effects (check specular_merge).
- [x] Verify Glow
- [x] Verify Auto exposure
- [x] Fix DOF

---

- *Production edit: This closes https://github.com/godotengine/godot-proposals/issues/4280.*

---

🍀 This work has been financed and kindly donated to the Godot Engine project by [W4 Games](https://w4games.com/). 🍀

mrjustaguy commented:
My initial report on FSR 2.2 vs TAA

FSR 2.2 is not compatible with MSAA, as it breaks rendering at lower resolutions, though it is usable with Native resolution if TAA is also enabled (else it turns into a jittering mess)

Aside From that, Comparing with TAA, There is no Motion pixelization with FSR2, However static scenes with high frequency detail are a little less stable, It also seems to be properly resolving Transparencies unlike TAA which breaks with transparencies causing anything affected by transparent surfaces to go crazy (say Quad effects show the underlying quad moving with TAA, not so with FSR 2.2)

FSR 2.2 costs 2x as TAA, However looks significantly better IMHO and is well worth it, but All other Anti Aliasing should probably be forced to off when FSR 2.2 is on

Saul2022 commented:
as noted by mrj quality increases a lot with fsr 2.2, though performance on an igpu is  pain xdd, though the flickering is gone unlike taa 1.0. some screenshots: at scaling quality at 0.5  the pixelated one is bilinear,  the flicker fsr 1 and the quality one is on 2.2![editor_screenshot_2023-08-31T175135](https://github.com/godotengine/godot/assets/97898580/5f0386de-4dd5-4723-b653-a787c8fcad78)![editor_screenshot_2023-08-31T175122](https://github.com/godotengine/godot/assets/97898580/4d5aef69-b18d-44a0-aab0-0393ec1f891c)
 
![editor_screenshot_2023-08-31T175038](https://github.com/godotengine/godot/assets/97898580/f92b7fe5-14f7-4c88-b3a8-0b73317ffe14)

DarioSamo commented:
> Note that this PR includes changes to optimize motion vector generation (by deriving from the depth buffer when possible), which may improve performance substantially, even with traditional native TAA. Therefore, this PR might be enough to close #61905 (I haven't tested yet).
We'll probably see the biggest gains when the render list is sorted in a way that only renders static geometry first and dynamic geometry afterwards. This PR doesn't take care of that nor does it add that behavior to the existing TAA shader, but we can very well take advantage of the implementation here.

> I'd recommend adding `// -- GODOT start --` and `// -- GODOT end --` before and after modified sections in FSR2 code and shaders, so that updating these files is easier in the future. We do this in third-party libraries where we need to modify the code.
Sounds good to me.

ettiSurreal commented:
Well someone has to ask about this.
FSR 3.0 is apparently supposed to come out as soon as this month, so any words on this?

Calinou commented:
> Well someone has to ask about this. FSR 3.0 is apparently supposed to come out as soon as this month, so any words on this?

The FSR 3 code isn't available yet. This will likely be worked on after it's open sourced, but we can't give an ETA.

BastiaanOlij commented on servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:
```diff
@@ -340,14 +340,16 @@ void RendererSceneRenderRD::_render_buffers_post_process_and_tonemap(const Rende
 	bool can_use_effects = target_size.x >= 8 && target_size.y >= 8; // FIXME I think this should check internal size, we do all our post processing at this size...
 	bool can_use_storage = _render_buffers_can_be_storage();
 
+	bool use_fsr = fsr && can_use_effects && rb->get_scaling_3d_mode() == RS::VIEWPORT_SCALING_3D_MODE_FSR;
```
I don't think we can do this any better until we have a proper render graph so we don't need to hardcode this stuff.

But seeing FSR is only embedded in the Forward+ renderer, once we start extending the generic code, this will start getting in the way pretty quickly. We've made for too many assumptions in the rendering code that are now no longer true and we really risk painting ourselves into a corner with more and more exceptions being embedded here as we introduce more effects into the rendering pipeline.

One alternative that may be nicer here but maybe a bit much as a change for FSR, is that we let go of the idea of internal texture as this is way too restrictive.

Instead at the start of rendering we determine our 3d render size (as a replacement of internal size) and we leave it up to `RendererSceneRenderRD::_render_scene` to do what it must do in order to render the 3D output, and it then simply returns the RID and size of whatever end result it has made, whether that is the original 3D render buffer (internal_texture) or the upscaled texture (FSR) or if an effect is run that copies the render output into a new texture, it no longer needs to do a copy back, it is that texture that becomes the output.

It also means that `RenderSceneBuffers` will no longer hardcode some of the buffers we use but again leave it up to the implementation of the render method to request/define required buffers, and in that we'll probably need to provide implementations like sky with the RID of the color texture it needs to render into, instead of it assuming that is the internal texture.

That would add way more flexibility into how we come to our rendering end result, and won't require hardcoding which buffers are being used and make it much easier to chain effects.

DarioSamo replied:
I agree with the general idea, although that kind of refactor may be out of my scope for now. Should we move it to a PR after this one?

BastiaanOlij replied:
Agreed

BastiaanOlij commented on servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:
```diff
@@ -567,12 +570,12 @@ void RendererSceneRenderRD::_render_buffers_post_process_and_tonemap(const Rende
 	}
 
-	if (fsr && can_use_effects && rb->get_scaling_3d_mode() == RS::VIEWPORT_SCALING_3D_MODE_FSR) {
+	if (use_fsr) {
```
btw, on a foot note, we should do this in more places in the code, instead of copying a complex check in multiple points, just set a variable as early on and use that variable. Makes code a lot easier to read :)

DarioSamo replied:
Yup, I'm a big fan of declaring boolean variables where the name is descriptive of what the condition means instead of throwing it all in the if clause. It tends to save on needing to have a comment describing what the condition is as well, and it should get optimized by the compiler fairly easily.

BastiaanOlij commented on servers/rendering/renderer_rd/renderer_scene_render_rd.cpp:
```diff
@@ -732,6 +735,11 @@ void RendererSceneRenderRD::_render_buffers_debug_draw(const RenderDataRD *p_ren
 		}
 	}
 
+	if (debug_draw == RS::VIEWPORT_DEBUG_DRAW_INTERNAL_BUFFER) {
```
Should probably have made this remark earlier on, but we need a better name then internal buffer. Internal buffer is far to non-descript. Which internal buffer? We have dozens of internal buffers...

DarioSamo replied:
Do you think we should solve this in this particular PR? It matches the name of the internal texture so that's pretty much why I used that.

Best alternatives I can think of is something like Color Buffer or related.

BastiaanOlij commented on servers/rendering/renderer_rd/shaders/effects/fsr2_accumulate_pass.glsl:
```diff
@@ -0,0 +1,8 @@
+#VERSION_DEFINES
+
+#include "motion_vector_inc.glsl"
+#include "thirdparty/amd-fsr2/shaders/ffx_fsr2_accumulate_pass.glsl"
```
nitpicking, needs a newline at the end, same goes for a number of the other files :)

DarioSamo replied:
Sounds good to me.

BastiaanOlij commented on servers/rendering/renderer_rd/storage_rd/render_scene_buffers_rd.cpp:
```diff
@@ -387,6 +398,10 @@ const RD::TextureFormat RenderSceneBuffersRD::get_texture_format(const StringNam
+	return get_texture_slice_view(p_context, p_texture_name, p_layer, p_mipmap, p_layers, p_mipmaps, RD::TextureView());
+}
+
+RID RenderSceneBuffersRD::get_texture_slice_view(const StringName &p_context, const StringName &p_texture_name, const uint32_t p_layer, const uint32_t p_mipmap, const uint32_t p_layers, const uint32_t p_mipmaps, RD::TextureView p_view) {
```
You also need to add a `_get_texture_slice_view` which uses `Ref<RDTextureView>` and add a bind method so this is accessible externally.

Have a look at `_create_texture_view` that does the same for `create_texture_view`.

DarioSamo replied:
Sounds fine to me, wasn't aware if we wanted a new API addition as I worked my way around modifying the existing one to keep compatibility.

DarioSamo replied:
Alright, we'll add the extra function.

BastiaanOlij replied:
Adding is fine, modifying is a problem :)

BastiaanOlij reviewed (commented):
Most of my remarks are more general ones, except for some of the FSR logic that ended up in `RenderSceneBuffers`, but either should be in `RenderBufferDataForwardClustered` or handled by the FSR effect itself like other effects do.

All in all this looks amazing!

Calinou commented:
This fails to build with MinGW from Linux (Fedora 38) for me:

```
Compiling servers/rendering/renderer_rd/storage_rd/utilities.cpp ...
servers/rendering/renderer_rd/effects/fsr2.cpp: In function 'FfxErrorCode create_pipeline_rd(FfxFsr2Interface*, FfxFsr2Pass, const FfxPipelineDescription*, FfxPipelineState*)':
servers/rendering/renderer_rd/effects/fsr2.cpp:334:47: error: invalid conversion from 'const wchar_t*' to 'size_t' {aka 'long long unsigned int'} [-fpermissive]
  334 |                         wcsncpy(binding.name, L"r_dilated_motion_vectors", sizeof(binding.name) / sizeof(wchar_t));
      |                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                               |
      |                                               const wchar_t*
servers/rendering/renderer_rd/effects/fsr2.cpp:334:97: error: invalid conversion from 'long long unsigned int' to 'const wchar_t*' [-fpermissive]
  334 |                         wcsncpy(binding.name, L"r_dilated_motion_vectors", sizeof(binding.name) / sizeof(wchar_t));
      |                                                                            ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
      |                                                                                                 |
      |                                                                                                 long long unsigned int
servers/rendering/renderer_rd/effects/fsr2.cpp:334:32: error: too few arguments to function 'errno_t wcsncpy_s(wchar_t*, size_t, const wchar_t*, size_t)'
  334 |                         wcsncpy(binding.name, L"r_dilated_motion_vectors", sizeof(binding.name) / sizeof(wchar_t));
      |                                ^
In file included from /usr/x86_64-w64-mingw32/sys-root/mingw/include/wchar.h:1561,
                 from /usr/x86_64-w64-mingw32/sys-root/mingw/include/c++/cwchar:44,
                 from /usr/x86_64-w64-mingw32/sys-root/mingw/include/c++/bits/postypes.h:40,
                 from /usr/x86_64-w64-mingw32/sys-root/mingw/include/c++/iosfwd:40,
                 from /usr/x86_64-w64-mingw32/sys-root/mingw/include/c++/system_error:40,
                 from /usr/x86_64-w64-mingw32/sys-root/mingw/include/c++/mutex:41,
                 from ./core/os/mutex.h:37,
                 from ./servers/rendering/renderer_rd/shader_rd.h:34,
                 from ./servers/rendering/renderer_rd/shaders/effects/fsr2_accumulate_pass.glsl.gen.h:5,
                 from servers/rendering/renderer_rd/effects/fsr2.h:34,
                 from servers/rendering/renderer_rd/effects/fsr2.cpp:31:
/usr/x86_64-w64-mingw32/sys-root/mingw/include/sec_api/wchar_s.h:316:27: note: declared here
  316 |   _CRTIMP errno_t __cdecl wcsncpy_s(wchar_t *_Dst,size_t _DstSizeInChars,const wchar_t *_Src,size_t _MaxCount);
      |                           ^~~~~~~~~
servers/rendering/renderer_rd/effects/fsr2.cpp: In function 'FfxResource get_resource_rd(RID*, const wchar_t*)':
servers/rendering/renderer_rd/effects/fsr2.cpp:496:27: error: invalid conversion from 'const wchar_t*' to 'size_t' {aka 'long long unsigned int'} [-fpermissive]
  496 |         wcsncpy(res.name, name, sizeof(res.name) / sizeof(wchar_t));
      |                           ^~~~
      |                           |
      |                           const wchar_t*
servers/rendering/renderer_rd/effects/fsr2.cpp:496:50: error: invalid conversion from 'long long unsigned int' to 'const wchar_t*' [-fpermissive]
  496 |         wcsncpy(res.name, name, sizeof(res.name) / sizeof(wchar_t));
      |                                 ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
      |                                                  |
      |                                                  long long unsigned int
servers/rendering/renderer_rd/effects/fsr2.cpp:496:16: error: too few arguments to function 'errno_t wcsncpy_s(wchar_t*, size_t, const wchar_t*, size_t)'
  496 |         wcsncpy(res.name, name, sizeof(res.name) / sizeof(wchar_t));
      |                ^
/usr/x86_64-w64-mingw32/sys-root/mingw/include/sec_api/wchar_s.h:316:27: note: declared here
  316 |   _CRTIMP errno_t __cdecl wcsncpy_s(wchar_t *_Dst,size_t _DstSizeInChars,const wchar_t *_Src,size_t _MaxCount);
      |                           ^~~~~~~~~
scons: *** [servers/rendering/renderer_rd/effects/fsr2.windows.editor.x86_64.o] Error 1
scons: building terminated because of errors.
[Time elapsed: 00:01:15.512]
```

Command line: `scons -j1 platform=windows optimize=speed lto=full progress=no debug_symbols=yes`

mrjustaguy commented:
there are some that don't look smooth and exhibit very similar FSR 2 behaviors, such as Star Wars Jedi Survivor mentioned above, though that is the only FSR 2 game where I've heard of such complaints so could be just bad implementation there too, and not an actual FSR 2 issue.

DarioSamo commented on servers/rendering/renderer_rd/environment/sky.cpp:
```diff
@@ -973,7 +973,7 @@ SkyRD::~SkyRD() {
 }
 
-void SkyRD::setup_sky(RID p_env, Ref<RenderSceneBuffersRD> p_render_buffers, const PagedArray<RID> &p_lights, RID p_camera_attributes, uint32_t p_view_count, const Projection *p_view_projections, const Vector3 *p_view_eye_offsets, const Transform3D &p_cam_transform, const Projection &p_cam_projection, const Size2i p_screen_size, RendererSceneRenderRD *p_scene_render) {
+void SkyRD::setup_sky(RID p_env, Ref<RenderSceneBuffersRD> p_render_buffers, const PagedArray<RID> &p_lights, RID p_camera_attributes, uint32_t p_view_count, const Projection *p_view_projections, const Vector3 *p_view_eye_offsets, const Transform3D &p_cam_transform, const Projection &p_cam_projection, const Size2i p_screen_size, Vector2 p_jitter, RendererSceneRenderRD *p_scene_render) {
```
It's used in the function for adding a jitter to the sky rendering. It helps with the accumulation process on lower resolutions, as it'll look visibly lower res otherwise.

BastiaanOlij replied:
Ah my bad, I missed that somehow. All good! :)

DarioSamo commented:
> @DarioSamo I've noticed the disoccluded pixels look pin-sharp, almost like the image is using nearest-neighbor filtering before being upscaled. Is this expected? I recall most games looking softer with newly disoccluded pixels, even if fizzle does occur in more complex scenes.

That sounds about right to me. You can check the sample in [https://github.com/GPUOpen-Effects/FidelityFX-FSR2](https://github.com/GPUOpen-Effects/FidelityFX-FSR2) as well if you'd like a reference to compare to from AMD themselves, but I don't think anything in the implementation could cause that.

DarioSamo commented on servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp:
```diff
@@ -1792,9 +1807,9 @@ void RenderForwardClustered::_render_scene(RenderDataRD *p_render_data, const Co
+				sky.setup_sky(p_render_data->environment, rb, *p_render_data->lights, p_render_data->camera_attributes, 1, &projection, &eye_offset, p_render_data->scene_data->cam_transform, projection, screen_size, Vector2(0.0f, 0.0f), this);
 			} else {
-				sky.setup_sky(p_render_data->environment, rb, *p_render_data->lights, p_render_data->camera_attributes, p_render_data->scene_data->view_count, p_render_data->scene_data->view_projection, p_render_data->scene_data->view_eye_offset, p_render_data->scene_data->cam_transform, p_render_data->scene_data->cam_projection, screen_size, this);
+				sky.setup_sky(p_render_data->environment, rb, *p_render_data->lights, p_render_data->camera_attributes, p_render_data->scene_data->view_count, p_render_data->scene_data->view_projection, p_render_data->scene_data->view_eye_offset, p_render_data->scene_data->cam_transform, p_render_data->scene_data->cam_projection, screen_size, p_render_data->scene_data->taa_jitter, this);
```
Yeah this line is starting to make a fair argument for the clang formatting to limit the width. :)

BastiaanOlij replied:
I introduced render data long after the sky implementation was made, it's whole purpose was to make it easy to have access to the state of rendering a frame throughout the various effects. Though I do agree it makes it harder to tell what bits of data each effect actually uses.

clayjohn replied:
I think a follow up PR is fine. This PR is already getting a bit too big

DarioSamo commented on servers/rendering/renderer_rd/storage_rd/render_scene_buffers_rd.cpp:
```diff
@@ -593,6 +614,24 @@ RID RenderSceneBuffersRD::get_depth_texture(const uint32_t p_layer) {
 
+// FSR2.
+
+void RenderSceneBuffersRD::ensure_fsr2(RendererRD::FSR2Effect *effect) {
```
Sounds good to me, I wasn't aware there was an specific render buffer storage class for Forward+. I'll do this change.

Saul2022 commented:
> 
> We'll probably see the biggest gains when the render list is sorted in a way that only renders static geometry first and dynamic geometry afterwards. 
 Yea though if you plan to add it will there be a sething where we choose if an object is static or dynamic ?, kind of similar to bastian pr, but for geometry.

DarioSamo commented:
> Yea though if you plan to add it will there be a sething where we choose if an object is static or dynamic ?, kind of similar to bastian pr, but for geometry.

Nope, the plan would be to auto-detect it.

Saul2022 commented:
> Nope, the plan would be to auto-detect it based on whether the element has moved or not.
Alright, but it can have an effect even dissabling taa?

DarioSamo commented:
No, there would be no effect when TAA or FSR2 isn't used as the performance degradation comes from writing motion vector information. If motion vectors aren't required, then there's no point in splitting the render lists.

DarioSamo commented:
I've added the optimization we talked about, a new type of render list for geometry that requires motion vectors will be used when FSR2 is used.

This improvement could be extended to the TAA shader once it's able to derive motion vectors from depth, but that'll probably be best separated into its own PR as there are some design changes that are required on the algorithm to support it correctly.

DarioSamo commented:
Also fixed compatibility with screenspace effects such as SSS.

I think this is ready for reviewing in further detail.

clayjohn commented on servers/rendering/renderer_rd/shaders/effects/fsr2_accumulate_pass.glsl:
```diff
@@ -0,0 +1,8 @@
+#[compute]
+
```
Let's move these fsr2 shader files into their own subfolder

clayjohn commented on servers/rendering/renderer_viewport.cpp:
```diff
@@ -151,10 +164,14 @@ void RendererViewport::_configure_3d_render_buffers(Viewport *p_viewport) {
 					render_width = MAX(width * scaling_3d_scale, 1.0); // width / (width * scaling)
 					render_height = MAX(height * scaling_3d_scale, 1.0);
+
+					// Implementation has been copied from ffxFsr2GetJitterPhaseCount.
+					jitter_phase_count = uint32_t(8.0f * pow(float(width) / render_width, 2.0f));
```
This adds jitter to regular FSR. Since regular FSR doesn't have a resolve pass it is super noticable. We need to force this to 0 with FSR1

clayjohn commented on servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp:
```diff
@@ -1889,40 +1943,74 @@ void RenderForwardClustered::_render_scene(RenderDataRD *p_render_data, const Co
 			RD::get_singleton()->draw_list_end();
 		}
+
+		if (will_continue_color && using_reactive) {
```
This should probably only run if motion vectors were used and the motion vector buffer is still open
```suggestion
		if (will_continue_color && using_reactive && render_motion_pass) {
```

DarioSamo replied:
That'd probably introduce an error because above this the motion vector buffer is cleared regardless of the render list having anything (which is what should happen as all pixels would be derived from camera motion).

DarioSamo replied:
Agreed. I'd suggest moving this inside the branch for the render motion pass instead.

clayjohn reviewed (commented):
Looks really good! I left a few comments and we also need to discuss the project settings. 

If FSR2.2 is a valid TAA option then we need to do more to communicate to the user that it shouldn't be used with TAA. Right now you can enable FSR2.2 and TAA at the same time.   But it clearly just uses FSR2.2

DarioSamo commented:
- Took the jitter out of FSR1.
- Simplified the motion vector buffer logic to be closed only when the motion pass is rendered, as it'll be cleared and closed early otherwise.
- Moved the FSR2 proxy shaders to a subfolder.

DarioSamo commented:
> If FSR2.2 is a valid TAA option then we need to do more to communicate to the user that it shouldn't be used with TAA. Right now you can enable FSR2.2 and TAA at the same time. But it clearly just uses FSR2.2

I forgot to mention the current approach shows a warning on the console similar to the rest of the behavior when upscalers are used with invalid configurations. Do you think we need something better?

This is the current handling:
```
if (use_taa && scaling_3d_mode == RS::VIEWPORT_SCALING_3D_MODE_FSR2) {
	// FSR2 can't be used with TAA.
	// Turn it off and prefer using FSR2.
	WARN_PRINT_ONCE("FSR 2 is not compatible with TAA. Turning TAA off.");
	use_taa = false;
}
```

AThousandShips commented on servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp:
```diff
@@ -105,6 +105,12 @@ void RenderForwardClustered::RenderBufferDataForwardClustered::ensure_voxelgi()
 	}
 }
 
+void RenderForwardClustered::RenderBufferDataForwardClustered::ensure_fsr2(RendererRD::FSR2Effect *effect) {
```
```suggestion
void RenderForwardClustered::RenderBufferDataForwardClustered::ensure_fsr2(RendererRD::FSR2Effect *p_effect) {
```

AThousandShips commented on servers/rendering/renderer_rd/shaders/effects/fsr2/SCsub:
```diff
@@ -0,0 +1,17 @@
+Import("env")
+
+if "RD_GLSL" in env["BUILDERS"]:
+    # find all include files
```
```suggestion
    # Find all include files.
```

DarioSamo replied:
I can fix the comment style here, but it's worth pointing out I just copied and pasted this file from the other folders that have the same build system: https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/shaders/environment/SCsub

Personally I'd prefer if all of this was handled by the parent script in a recursive way, but that's how the current architecture works. It might harm the possibility of finding and replacing all of the instances of the file later if I make it different here. Should I still go ahead with changing it?

akien-mga replied:
I would prefer keeping it consistent, even if it doesn't fit the style.

This can be refactored later all at once (include possibly reducing the code duplication).

AThousandShips commented on servers/rendering/renderer_rd/shaders/effects/fsr2/SCsub:
```diff
@@ -0,0 +1,17 @@
+    # find all shader code(all glsl files excluding our include files)
+    glsl_files = [str(f) for f in Glob("*.glsl") if str(f) not in gl_include_files]
+
+    # make sure we recompile shaders if include files change
```
```suggestion
    # Make sure we recompile shaders if include files change.
```

AThousandShips commented on servers/rendering/renderer_rd/shaders/effects/fsr2/SCsub:
```diff
@@ -0,0 +1,17 @@
+    # find all include files
+    gl_include_files = [str(f) for f in Glob("*_inc.glsl")] + [str(f) for f in Glob("../*_inc.glsl")]
+
+    # find all shader code(all glsl files excluding our include files)
```
```suggestion
    # Find all shader code(all glsl files excluding our include files).
```

akien-mga commented on servers/rendering/renderer_rd/effects/SCsub:
```diff
@@ -2,4 +2,33 @@
-env.add_source_files(env.servers_sources, "*.cpp")
+env_effects = env.Clone()
+
+# Thirdparty source files
```
That comment is copied from dozen other SCsub files, it's better to keep it as is for consistency.

AThousandShips commented on servers/rendering/renderer_scene_cull.cpp:
```diff
@@ -37,6 +37,19 @@
+		index = index / base;
+	}
+	return r * 2.0f - 1.0f;
+};
```
```suggestion
}
```

AThousandShips commented on servers/rendering/renderer_scene_cull.cpp:
```diff
@@ -37,6 +37,19 @@
 #include <new>
 
+/* HALTON SEQUENCE */
+
```
```suggestion

#ifndef _3D_DISABLED
```

AThousandShips commented on servers/rendering/renderer_scene_cull.cpp:
```diff
@@ -37,6 +37,19 @@
+	}
+	return r * 2.0f - 1.0f;
+}
+
```
```suggestion
#endif // _3D_DISABLED

```

AThousandShips commented on servers/rendering/renderer_rd/storage_rd/mesh_storage.cpp:
```diff
@@ -1392,12 +1392,18 @@ void MeshStorage::_multimesh_get_motion_vectors_offsets(RID p_multimesh, uint32_
 
+bool MeshStorage::_multimesh_uses_motion_vectors_offsets(RID p_multimesh) {
+	MultiMesh *multimesh = multimesh_owner.get_or_null(p_multimesh);
+	ERR_FAIL_COND_V(!multimesh, false);
```
```suggestion
	ERR_FAIL_NULL_V(multimesh, false);
```

AThousandShips commented on servers/rendering/renderer_rd/effects/fsr2.cpp:
```diff
@@ -0,0 +1,889 @@
+}
+
+static FfxErrorCode schedule_gpu_job_rd(FfxFsr2Interface *p_backend_interface, const FfxGpuJobDescription *p_job) {
+	ERR_FAIL_COND_V(!p_backend_interface, FFX_ERROR_INVALID_ARGUMENT);
```
```suggestion
	ERR_FAIL_NULL_V(p_backend_interface, FFX_ERROR_INVALID_ARGUMENT);
```

AThousandShips commented on servers/rendering/renderer_rd/effects/fsr2.cpp:
```diff
@@ -0,0 +1,889 @@
+
+static FfxErrorCode schedule_gpu_job_rd(FfxFsr2Interface *p_backend_interface, const FfxGpuJobDescription *p_job) {
+	ERR_FAIL_COND_V(!p_backend_interface, FFX_ERROR_INVALID_ARGUMENT);
+	ERR_FAIL_COND_V(!p_job, FFX_ERROR_INVALID_ARGUMENT);
```
```suggestion
	ERR_FAIL_NULL_V(p_job, FFX_ERROR_INVALID_ARGUMENT);
```

AThousandShips commented on servers/rendering/renderer_rd/effects/fsr2.cpp:
```diff
@@ -0,0 +1,889 @@
+}
+
+static FfxErrorCode execute_gpu_jobs_rd(FfxFsr2Interface *p_backend_interface, FfxCommandList p_command_list) {
+	ERR_FAIL_COND_V(!p_backend_interface, FFX_ERROR_INVALID_ARGUMENT);
```
```suggestion
	ERR_FAIL_NULL_V(p_backend_interface, FFX_ERROR_INVALID_ARGUMENT);
```

mrjustaguy commented:
I think I've noticed some issues with FSR 2 implementation in Godot.

Glow and Auto exposure seem to be applied Before FSR, not After, Likewise DOF doesn't work at all with FSR 2 turned on - https://gpuopen.com/manuals/fidelityfx_sdk/fidelityfx_sdk-page_techniques_super-resolution-temporal/#algorithm-structure

DarioSamo commented:
> I think I've noticed some issues with FSR 2 implementation in Godot.
> 
> Glow and Auto exposure seem to be applied Before FSR (causing significant visual differences between native and non native res), not After, Likewise DOF doesn't work at all with FSR 2 turned on - https://gpuopen.com/manuals/fidelityfx_sdk/fidelityfx_sdk-page_techniques_super-resolution-temporal/#algorithm-structure

FWIW, FSR2 is applied pretty much on the same step where the current TAA is, so the exposure texture that is used as the input is pretty much the value from the previous frame because the exposure is calculated after the TAA. If it's deemed as necessary it might be needed to shuffle around the steps a bit depending on whether FSR2 is used or not I guess, but I don't think FSR2 could be moved to be after the current exposure step as it'd break other things.

Regardless, those are some pretty interesting points to check. I'll add them to the TODO.

Rytelier commented:
There appear to be ghosting related to moving shaders which probably should be investigated.
I tested it on a "mirror" camera viewport shader which has ghosting when moving shader's camera.
![image](https://github.com/godotengine/godot/assets/45795134/95791ad4-ae83-41ee-855a-ca479f5fbb05)

DarioSamo commented:
> There appear to be ghosting related to moving shaders which probably should be investigated. I tested it on a "mirror" camera viewport shader which has ghosting when moving shader's camera. ![image](https://user-images.githubusercontent.com/45795134/268308237-95791ad4-ae83-41ee-855a-ca479f5fbb05.png)

I'm not sure there's much to investigate here, it's not gonna grab motion vectors from an animated surface.

The best we can hope for is to allow people to be able to output to the reactivity mask somehow, but that's not a feature that Godot currently has as a render target. I'm fairly sure you'd find the same problems with TAA as well.

DarioSamo commented:
I've verified DOF is broken when using FSR2 as it was suggested on a comment before, will look into fixing it.

DarioSamo commented:
@mrjustaguy DoF should be working now, thanks for the report.

Calinou commented:
I've tested performance again on https://github.com/Calinou/godot-reflection/tree/fsr2-test with the latest revision of this PR (cf3fd3b1a). All tests done in 3840×2160 on a GeForce RTX 4090.

- **No AA:** 334 FPS
- **Native TAA:** 264 FPS
- **FSR2 native:** 227 FPS
- **FSR2 quality:** 358 FPS
- **FSR2 balanced:** 405 FPS
- **FSR2 performance:** 459 FPS

PS: Is https://github.com/godotengine/godot/pull/81350 incorporated in this PR? I've tested the latest revision of this PR, but mipmap bias isn't adjusted according to AMD's formula when FSR2 is enabled. This is noticeable if you decrease resolution scale all the way down near a checkerboard texture.

InitialCon commented:
Testing out this PR with Screen Space Reflections enabled and it seems that using FSR2 causes them to be even more jittery than when using the built-in taa. Is this a bug or just a downside of them being less blurry? 
![SSRBugFSR2](https://github.com/godotengine/godot/assets/66960524/ee60fa16-41b1-44df-a98a-3b04068ed7fa)
I'm using @Calinou's reflection test scene with normal settings.

Calinou commented:
> Is this a bug or just a downside of them being less blurry?

I think this is just a consequence of them looking less blurry. [SSR needs to be redesigned to take advantage of TAA jitter into account at some point.](https://github.com/godotengine/godot/issues/70526)

InitialCon commented:
Cool, just wanted to be sure in case it was a bug. I've seen the sdfgi lighting jittering a bit too but that might also be a downside of using a sharper taa solution.

DarioSamo commented:
> PS: Is #81350 incorporated in this PR? I've tested the latest revision of this PR, but mipmap bias isn't adjusted according to AMD's formula when FSR2 is enabled. This is noticeable if you decrease resolution scale all the way down near a checkerboard texture. Further manual changes of the **Texture Mipmap Bias** project settings aren't visible either.

Nope it's not incorporated. Now that it's merged I'll rebase it based on that (although I'll go a bit less aggressive than AMD's formula by using -0.5 instead of -1 as it causes some visible jittering)

DarioSamo committed:
Add FidelityFX Super Resolution 2.2 (FSR 2.2.1) support.

Introduces support for FSR2 as a new upscaler option available from the project settings. Also introduces an specific render list for surfaces that require motion and the ability to derive motion vectors from depth buffer and camera motion.

DarioSamo commented:
Rebased and now mipmap bias works properly. Texture quality should now be maintained more properly when using lower scales.

mrjustaguy commented:
I don't remember if this was mentioned anywhere before, so here goes:
What's the plan regarding Mesh LODing with FSR2 on? Are they planned to be always running as Native LODs with FSR2 or using current behavior (which shows Lower LODs in the reconstruction that can become quite apparent, not so with Bilinear/FSR1)

clayjohn reviewed (commented):
Looks great to me! I left one nitpick on an error message, but overall the code looks very good. 

I tested locally with the TPS demo and tried various post FX (DoF, Glow, Auto exposure, etc) and it appears to work well. 

Let's get this merged before the next dev release so we can get some testing before beta.

clayjohn reviewed (approved):
Should be good to merge after a rebase to squash the commits

DarioSamo commented:
Done.

bruvzg commented on thirdparty/README.md:
```diff
@@ -16,6 +16,19 @@ Files extracted from upstream source:
 - `ffx_a.h` and `ffx_fsr1.h` from `ffx-fsr`
 - `license.txt`
 
+## amd-fsr2
```
Probably license info should be added to https://github.com/godotengine/godot/blob/master/COPYRIGHT.txt as well, after fsr-1:

https://github.com/godotengine/godot/blob/df0a822323a79e1a645f0c6a17d51c7602f23166/COPYRIGHT.txt#L139-L142

DarioSamo replied:
Gonna delegate to @akien-mga to give me the info on this one just in case.

akien-mga commented on thirdparty/README.md:
```diff
@@ -16,6 +16,19 @@ Files extracted from upstream source:
 - `ffx_a.h` and `ffx_fsr1.h` from `ffx-fsr`
 - `license.txt`
 
+## amd-fsr2
```
We keep two lines between each section in this file (but I've learned over time that programmers seem to be pretty oblivious of that detail :)). So add one line before, and after.

DarioSamo replied:
Done.

akien-mga reviewed (approved):
Third-party code integration looks good.

akien-mga commented:
Thanks!
